lib/web/cachehandler.go

```diff
@@ -20,15 +20,27 @@ package web
 
 import (
 	"net/http"
+	"path/filepath"
 	"time"
 
 	"github.com/gravitational/teleport/lib/httplib"
 )
 
 // makeCacheHandler adds support for gzip compression for given handler.
-func makeCacheHandler(handler http.Handler) http.Handler {
+func makeCacheHandler(handler http.Handler, etag string) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		httplib.SetCacheHeaders(w.Header(), time.Hour*24*365 /* one year */)
+		// for javascript assets, we want to set an entity tag based on the version
+		// of teleport. This will allow us to cache our "app.js", and validate if the file
+		// should be recached based on the teleport version. A webasset will not change within
+		// the same version so we can use this etag to return a 304 (Not Modified). This results in
+		// only a couple hundred bytes response compared to the entire bundle js file
+
+		// The rest of our assets like fonts can be cached between versions without issue
+		if filepath.Ext(r.URL.Path) == ".js" {
+			httplib.SetEntityTagCacheHeaders(w.Header(), etag)
+		} else {
+			httplib.SetCacheHeaders(w.Header(), time.Hour*24*365 /* one year */)
+		}
```

Comment:
If none of our resources are changing names between versions now (correct me if I'm wrong), then let's just use the same ETag mechanism for all of them.

Reply:
Definitely an option. I kept the old "cache forever" type of mechanism for other webassets because they are mostly fonts and I don't think we'd ever really need to revalidate those anyway so it just saves an extra network request. I won't die on that hill tho

Reply:
What about css?

Reply:
We could do `js`, `css`, and `wasm` here and let the rest cache forever. i don't think svg and png will change much (at least, ive never changed an svg and kept the name. usually its a _different_ svg from my anecdotal evidence). And fonts themselves won't ever change.

Or we can make it simple and reverse the logic and say "everything is etag except `woff`, `woff2`, and `ttf`. I'm a fan of the first more but would love your input.

Reply:
Point is that we have to update `css` (and `wasm`, good catch!) along with `js`, for correctness. I prefer the latter option where we default to correctness, and only optimize on specific files/types.

Reply:
Roger. Will do the ETag on all and only optimize fonts. If we find something later can be optimized, we'll add it then. Appreciate the feedback

Reply:
Updated. Looks like after the initial fetch, our "homepage" sends only ~3kb of webasset data (stuff that isn't fetched json like `resources`). and nearly all of _that_ 3k is the index.html page (1.8k suprises me but lets move on).

The stuff that does come over the wire is 800b of `config.js` which isn't really an asset and is used similar to a fetch request of info so, we can ignore that too

So about 250b for each "not modified" response so, `app.js`, `index.css`, and thats it.

Really what we are looking at is 500b and the rest is stuff that was already coming through that way before the change. not bad.